common/archiver/URI.go

```diff
@@ -0,0 +1,93 @@
+// Copyright (c) 2019 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package archiver
+
+import (
+	"errors"
+	"net/url"
+)
+
+type (
+	// URI identifies the scheme and location an archiver can archive or get history/visibility records
+	URI interface {
+		Scheme() string
+		Path() string
+		Hostname() string
+		Port() string
+		Username() string
+		Password() string
+		String() string
+	}
+
+	uri struct {
+		url *url.URL
+	}
+)
+
+// NewURI constructs a new archiver URI from string.
+func NewURI(s string) (URI, error) {
```

Comment:
Added tests. I returned RawPath in this case, as this URI will used by different implementations, I don't think it should do the escape and return the modified path. If an implementation need an escaped path, it can write its own logic to do so.